core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp

```diff
@@ -256,8 +265,10 @@ class Kokkos::Impl::ParallelFor<FunctorType, Kokkos::MDRangePolicy<Traits...>,
                 {global_x, global_y, global_z}, {local_x, local_y, local_z})
                 .exec_range();
           });
-        });
-    q.submit_barrier(std::vector<sycl::event>{parallel_for_event});
+    });
+    // This barrier would prevent q.memcpy for subsequent kernels from being
+    // brought forward in time.
+    // q.submit_barrier(std::vector<sycl::event>{parallel_for_event});
```

Comment:
I think this should be reverted in the end (at the moment it doesn't matter too much because we create queues as in-order as a workaround for other issues).